File: app/src/aether/tools/read_file.py
"""Read File tool — read a file's contents."""

from __future__ import annotations

import os

from aether.tools.base import AetherTool, ToolParam, ToolResult

MAX_FILE_SIZE = 100_000  # 100KB limit


class ReadFileTool(AetherTool):
    name = "read_file"
    description = "Read the contents of a file at the given path. Returns the file text."
    status_text = "Reading file..."
    parameters = [
        ToolParam(name="path", type="string", description="Absolute or relative file path to read"),
    ]

    def __init__(self, working_dir: str = "."):
        self.working_dir = os.path.abspath(working_dir)

    def _resolve(self, path: str) -> str:
        resolved = os.path.abspath(os.path.join(self.working_dir, path))
        if not resolved.startswith(self.working_dir):
            raise PermissionError(f"Path escapes working directory: {path}")
        return resolved

    async def execute(self, path: str) -> ToolResult:
        resolved = self._resolve(path)

        if not os.path.exists(resolved):
            return ToolResult.fail(f"File not found: {path}")

        if not os.path.isfile(resolved):
            return ToolResult.fail(f"Not a file: {path}")

        size = os.path.getsize(resolved)
        if size > MAX_FILE_SIZE:
            return ToolResult.fail(f"File too large: {size:,} bytes (max {MAX_FILE_SIZE:,})")

        try:
            with open(resolved, "r", encoding="utf-8", errors="replace") as f:
                content = f.read()
            return ToolResult.success(content, path=path, size=size)
        except Exception as e:
            return ToolResult.fail(f"Error reading file: {e}")

File: agent/internal/tools/builtin/read_file.go
package builtin

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/suryaumapathy/core-ai/agent/internal/tools"
)

type ReadFileTool struct{}

func (t *ReadFileTool) Definition() tools.Definition {
	return tools.Definition{
		Name:        "read_file",
		Description: "Read a text file from the working directory.",
		StatusText:  "Reading file...",
		Parameters: []tools.Param{
			{Name: "path", Type: "string", Description: "Relative file path", Required: true},
			{Name: "offset", Type: "integer", Description: "1-indexed line offset", Required: false, Default: 1},
			{Name: "limit", Type: "integer", Description: "Maximum lines to read", Required: false, Default: 200},
		},
	}
}

func (t *ReadFileTool) Execute(ctx context.Context, call tools.Call) tools.Result {
	_ = ctx
	path, _ := call.Args["path"].(string)
	offset, _ := toolsAsInt(call.Args["offset"])
	limit, _ := toolsAsInt(call.Args["limit"])
	if offset < 1 {
		offset = 1
	}
	if limit <= 0 {
		limit = 200
	}
	resolved, err := resolveSafePath(call.Ctx.WorkingDir, path)
	if err != nil {
		return tools.Fail("Invalid path: "+err.Error(), nil)
	}
	f, err := os.Open(resolved)
	if err != nil {
		return tools.Fail("Failed to open file: "+err.Error(), nil)
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	lineNo := 0
	count := 0
	lines := []string{}
	for scanner.Scan() {
		lineNo++
		if lineNo < offset {
			continue
		}
		lines = append(lines, fmt.Sprintf("%d: %s", lineNo, scanner.Text()))
		count++
		if count >= limit {
			break
		}
	}
	if err := scanner.Err(); err != nil {
		return tools.Fail("Failed reading file: "+err.Error(), nil)
	}
	return tools.Success(strings.Join(lines, "\n"), map[string]any{"path": resolved, "offset": offset, "limit": limit})
}

var _ tools.Tool = (*ReadFileTool)(nil)
